File: packages/mongoose/src/models/conversation.ts
import { IUser } from "@repo/shared-types";
import mongoose, { Schema, Document, Mongoose } from "mongoose";

export interface IConversation {
  participants: IUser[];
  isGroup: boolean;
  groupName?: string;
  groupAvatar?: string;
  // lastMessage?: {
  //   sender: mongoose.Types.ObjectId;
  //   text?: string;
  //   media?: string;
  //   timestamp: Date;
  // };
  lastMessage: mongoose.Types.ObjectId;
  unreadCount: Map<string, number>;
  isPinned: Map<string, boolean>;
  isMuted: Map<string, boolean>;
  archived: Map<string, boolean>;
  deletedFor: mongoose.Types.ObjectId[];
  typingUsers: mongoose.Types.ObjectId[];
  contactName: string;
  createdAt: Date;
  updatedAt: Date;
  pendingMessages: mongoose.Types.ObjectId[] | string[];
}

interface MConversation extends IConversation, Document {}

const ConversationSchema = new Schema<MConversation>(
  {
    participants: [
      { type: Schema.Types.ObjectId, ref: "User", required: true },
    ],
    isGroup: { type: Boolean, default: false },
    groupName: { type: String },
    groupAvatar: { type: String },
    lastMessage: { type: Schema.Types.ObjectId, ref: "Message" },
    unreadCount: { type: Map, of: Number, default: {} },
    isPinned: { type: Map, of: Boolean, default: {} },
    isMuted: { type: Map, of: Boolean, default: {} },
    archived: { type: Map, of: Boolean, default: {} },
    deletedFor: [{ type: Schema.Types.ObjectId, ref: "User" }],
    typingUsers: [{ type: Schema.Types.ObjectId, ref: "User" }],
    contactName: { type: String },
    pendingMessages: [{ type: Schema.Types.ObjectId, ref: "Message" }],
  },
  { timestamps: true },
);

export const ConversationModel = mongoose.model<MConversation>(
  "Conversation",
  ConversationSchema,
);

File: apps/matching-go/types/conversation.go
package types

import (
	"time"

	"go.mongodb.org/mongo-driver/v2/bson"
)

type LastMessage struct {
	Sender    bson.ObjectID `bson:"sender,omitempty" json:"sender,omitempty"`
	Text      string        `bson:"text,omitempty" json:"text,omitempty"`
	Media     string        `bson:"media,omitempty" json:"media,omitempty"`
	Timestamp time.Time     `bson:"timestamp" json:"timestamp"`
}

type Conversation struct {
	ID           bson.ObjectID   `bson:"_id,omitempty" json:"id,omitempty"`
	Participants []bson.ObjectID `bson:"participants" json:"participants"`
	IsGroup      bool            `bson:"isGroup" json:"isGroup"`
	GroupName    string          `bson:"groupName" json:"groupName"`
	GroupAvatar  string          `bson:"groupAvatar" json:"groupAvatar"`
	LastMessage  *LastMessage    `bson:"lastMessage" json:"lastMessage"`
	UnreadCount  map[string]int  `bson:"unreadCount" json:"unreadCount"`
	IsPinned     map[string]bool `bson:"isPinned" json:"isPinned"`
	IsMuted      map[string]bool `bson:"isMuted" json:"isMuted"`
	Archived     map[string]bool `bson:"archived" json:"archived"`
	DeletedFor   []bson.ObjectID `bson:"deletedFor" json:"deletedFor"`
	TypingUsers  []bson.ObjectID `bson:"typingUsers" json:"typingUsers"`
	ContactName  string          `bson:"contactName" json:"contactName"`
	CreatedAt    time.Time       `bson:"createdAt" json:"createdAt"`
	UpdatedAt    time.Time       `bson:"updatedAt" json:"updatedAt"`
}
